### PR TITLE
Added WeightModifierType.DISCRETIZE_PER_CHANNEL

### PR DESCRIPTION
**Pull request by jubueche**
## Related issues

No related issue for this new feature, but in this PR, I also want to address [this issue](https://github.com/IBM/aihwkit/issues/614).

## Description

I added a new `WeightModifierType.DISCRETIZE_PER_CHANNEL` type.

## Details

So far, we only have `WeightModifierType.DISCRETIZE` which discretizes the weight per-tensor. However, that induces too much quantization noise, so people resort to per-channel discretization, which is what I have implemented.
I also implemented a test that checks the correctness against a by-hand quantization in pure torch.

### CODE REVIEW

**Comment by jubueche** [comment]
@maljoras the implementation already works for the CPU part, but I still need to implement it in CUDA for the GPU.
As written in the description, we can also use this to update the design of specifying res for this modifier.
@kkvtran maybe you can help with updating the documentation once the technical part is done and reviewed by @maljoras .

File: src/rpucuda/cuda/weight_modifier_cuda.cu [hunk 1]
```diff
@@ -425,12 +495,19 @@ void WeightModifierCuda<T>::apply(
...
   T *amax = nullptr;
   if (wmpar.rel_to_actual_wmax && wmpar.type != WeightModifierType::Copy) {
-    if (!amaximizer_) {
-      amaximizer_ = RPU::make_unique<Maximizer<T>>(
-          context_, wmpar.copy_last_column ? (size_ - d_size_) : size_, true);
+    if (wmpar.type == WeightModifierType::DiscretizePerChannel) {
+      cudaMalloc((void**)&amax, d_size_ * sizeof(T));
+      int threadsPerBlock = 256; // what is better for this?
+      int blocksPerGrid = (x_size_ * d_size_ + threadsPerBlock - 1) / threadsPerBlock;
+      maxAbsAlongDimension<<<blocksPerGrid, threadsPerBlock, threadsPerBlock * sizeof(T)>>>(weights, x_size_, d_size_, amax, 0);
```
Comment (jubueche) [inline src/rpucuda/cuda/weight_modifier_cuda.cu:502]: I compute the max(abs(.)) per column here

File: src/rpucuda/cuda/weight_modifier_cuda.cu [hunk 2]
```diff
@@ -462,6 +539,17 @@ void WeightModifierCuda<T>::apply(
     }
     break;
   }
+  case WeightModifierType::DiscretizePerChannel: {
+    if (wmpar.res > (T)0.0) {
+      kernelModifyWeightsDiscretizePerChannel<T><<<nblocks, nthreads, 0, s>>>(
+          size_, d_size_, wmpar.copy_last_column, new_weights, weights, wmpar.res, wmpar.sto_round,
+          wmpar.assumed_wmax, amax,
+          wmpar.sto_round ? context_->getRandomStates(nblocks * nthreads) : nullptr);
+      cudaFree(amax);
```
Comment (jubueche) [inline src/rpucuda/cuda/weight_modifier_cuda.cu:548]: I do the discretization and free the memory again that was used for the per-column amax.

File: tests/test_cuda_discretize_per_channel.py
Comment (jubueche) [inline tests/test_cuda_discretize_per_channel.py:1]: This script passes, but the test does not pass. The test has some weights that are zero.

Comment (jubueche) [reply tests/test_cuda_discretize_per_channel.py:1]: @maljoras after the weight modification of discretize per column, some weights are zero and therefore my test is not passing. This has something to do with CUDA, but I don't know how to fix it. Do you have an idea?

**Comment by maljoras** [comment]
Is this really needed? That was the reason of having the weight remapping. So essentially, when you use channel wise weight remapping, you will get channel wise discretion with the original weight modifier for free

**Comment by jubueche** [comment]
You're right, I forgot about the channel-wise re-mapping.

**Pull request closed without merging**